Userland/Libraries/LibC/pwd.cpp

```diff
@@ -134,123 +128,61 @@ int getpwent_r(struct passwd* passwd_buf, char* buffer, size_t buffer_size, stru
+            *passwd_entry_ptr = nullptr;
+            return ENOENT;
         }
 
         if (ferror(s_stream)) {
-            dbgln("getpwent(): Read error: {}", strerror(ferror(s_stream)));
-            errno = EIO;
-            return -1;
+            *passwd_entry_ptr = nullptr;
+            return ferror(s_stream);
         }
 
         ++s_line_number;
         char* s = fgets(buffer, buffer_size, s_stream);
 
-        // Silently tolerate an empty line at the end.
         if ((!s || !s[0]) && feof(s_stream)) {
             *passwd_entry_ptr = nullptr;
-            return 0;
+            return ENOENT;
         }
 
         if (strlen(s) == buffer_size - 1) {
-            errno = ERANGE;
-            return -1;
+            *passwd_entry_ptr = nullptr;
+            return ERANGE;
         }
 
         if (parse_pwddb_entry(buffer, *passwd_buf)) {
             *passwd_entry_ptr = passwd_buf;
             return 0;
         }
-        // Otherwise, proceed to the next line.
     }
 }
 
-static void construct_pwd(struct passwd* pwd, char* buf, struct passwd** result)
-{
-    auto* buf_name = &buf[0];
-    auto* buf_passwd = &buf[s_name.length() + 1];
-    auto* buf_gecos = &buf[s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_dir = &buf[s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_shell = &buf[s_dir.length() + 1 + s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-
-    bool ok = true;
-    ok = ok && s_name.copy_characters_to_buffer(buf_name, s_name.length() + 1);
-    ok = ok && s_passwd.copy_characters_to_buffer(buf_passwd, s_passwd.length() + 1);
-    ok = ok && s_gecos.copy_characters_to_buffer(buf_gecos, s_gecos.length() + 1);
-    ok = ok && s_dir.copy_characters_to_buffer(buf_dir, s_dir.length() + 1);
-    ok = ok && s_shell.copy_characters_to_buffer(buf_shell, s_shell.length() + 1);
-
-    VERIFY(ok);
-
-    *result = pwd;
-    pwd->pw_name = buf_name;
-    pwd->pw_passwd = buf_passwd;
-    pwd->pw_gecos = buf_gecos;
-    pwd->pw_dir = buf_dir;
-    pwd->pw_shell = buf_shell;
-}
-
-int getpwnam_r(char const* name, struct passwd* pwd, char* buf, size_t buflen, struct passwd** result)
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwnam.html
+int getpwnam_r(char const* name, struct passwd* pwd, char* buffer, size_t bufsize, struct passwd** result)
 {
-    // FIXME: This is a HACK!
-    TemporaryChange name_change { s_name, {} };
-    TemporaryChange passwd_change { s_passwd, {} };
-    TemporaryChange gecos_change { s_gecos, {} };
-    TemporaryChange dir_change { s_dir, {} };
-    TemporaryChange shell_change { s_shell, {} };
-
     setpwent();
-    bool found = false;
-    while (auto* pw = getpwent()) {
-        if (!strcmp(pw->pw_name, name)) {
-            found = true;
-            break;
+    for (;;) {
+        if (auto rc = getpwent_r(pwd, buffer, bufsize, result); rc != 0)
+            return rc;
+        if (!strcmp(pwd->pw_name, name)) {
```

Comment:
Nitpick: absolutely not a blocker, but to be friendly to new devs, could we say `strcmp() == 0` here, instead of `!strcmp()`? The latter reads (at least to me) as "if the strings are _not_ equal". I had to think for a moment before remembering that `strcmp` returns 0 on equality and that !0 == 1 == true, which new devs, especially ones that haven't worked with c-strings before, might not realize. :^)